src/lib/mdns/ServiceNaming.cpp

```diff
@@ -67,43 +48,34 @@ CHIP_ERROR ExtractIdFromInstanceName(const char * name, PeerId * peerId)
     ReturnErrorCodeIf(name == nullptr, CHIP_ERROR_INVALID_ARGUMENT);
     ReturnErrorCodeIf(peerId == nullptr, CHIP_ERROR_INVALID_ARGUMENT);
 
-    peerId->SetNodeId(0);
-    peerId->SetFabricId(0);
-
-    bool deliminatorFound = false;
-    bool hasFabricPart    = false;
-    bool hasNodePart      = false;
-
-    for (; *name != '\0'; name++)
-    {
-        if (*name == '.')
-        {
-            break;
-        }
-        else if (*name == '-')
-        {
-            deliminatorFound = true;
-            continue;
-        }
-
-        uint8_t val = HexToInt(*name);
-        ReturnErrorCodeIf(val == UINT8_MAX, CHIP_ERROR_WRONG_NODE_ID);
-
-        if (deliminatorFound)
-        {
-            hasNodePart = true;
-            peerId->SetNodeId(peerId->GetNodeId() * 16 + val);
-        }
-        else
-        {
-            hasFabricPart = true;
-            peerId->SetFabricId(peerId->GetFabricId() * 16 + val);
-        }
-    }
-
-    ReturnErrorCodeIf(!deliminatorFound, CHIP_ERROR_WRONG_NODE_ID);
-    ReturnErrorCodeIf(!hasNodePart, CHIP_ERROR_WRONG_NODE_ID);
-    ReturnErrorCodeIf(!hasFabricPart, CHIP_ERROR_WRONG_NODE_ID);
+    // Make sure the string is long enough.
+    static constexpr size_t fabricIdByteLength   = 8;
+    static constexpr size_t fabricIdStringLength = fabricIdByteLength * 2;
+    static constexpr size_t nodeIdByteLength     = 8;
+    static constexpr size_t nodeIdStringLength   = nodeIdByteLength * 2;
+    static constexpr size_t totalLength          = fabricIdStringLength + nodeIdStringLength + 1; // +1 for '-'
+    size_t len                                   = strlen(name);
+    ReturnErrorCodeIf(len < totalLength, CHIP_ERROR_INVALID_ARGUMENT);
+
+    // Check that we have a proper terminator.
+    ReturnErrorCodeIf(name[totalLength] != '\0' && name[totalLength] != '.', CHIP_ERROR_WRONG_NODE_ID);
+
+    // Check what we have a separator where we expect.
+    ReturnErrorCodeIf(name[fabricIdStringLength] != '-', CHIP_ERROR_WRONG_NODE_ID);
+
+    static constexpr size_t bufferSize = max(fabricIdByteLength, nodeIdByteLength);
+    uint8_t buf[bufferSize];
+
+    ReturnErrorCodeIf(Encoding::HexToBytes(name, fabricIdStringLength, buf, bufferSize) == 0, CHIP_ERROR_WRONG_NODE_ID);
+    // Buf now stores the fabric id, as big-endian bytes.
+    static_assert(fabricIdByteLength == sizeof(uint64_t), "Casting won't work");
+    peerId->SetFabricId(Encoding::BigEndian::HostSwap64(*reinterpret_cast<uint64_t *>(buf)));
```

Comment:
raw bytes and casting feels a bit unclean/tricky code. Consider extracting into a separate method.